cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,810 @@
+};
+
+/**
+ * @brief Constructs `d_json_column` from node tree representation
+ * Newly constructed columns are insert into `root`'s children.
+ * `root` must be a list type.
+ *
+ * @param input Input JSON string device data
+ * @param tree Node tree representation of the JSON string
+ * @param col_ids Column ids of the nodes in the tree
+ * @param row_offsets Row offsets of the nodes in the tree
+ * @param root Root node of the `d_json_column` tree
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the device memory
+ * of child_offets and validity members of `d_json_column`
+ */
+void make_json_column2(device_span<SymbolT const> input,
+                       tree_meta_t& tree,
+                       device_span<NodeIndexT> col_ids,
+                       device_span<size_type> row_offsets,
+                       d_json_column& root,
+                       rmm::cuda_stream_view stream,
+                       rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  // 1. gather column information.
+  auto [d_column_tree, d_unique_col_ids, d_max_row_offsets] =
+    reduce_to_column_tree(tree, col_ids, row_offsets, stream);
+  auto num_columns    = d_unique_col_ids.size();
+  auto unique_col_ids = cudf::detail::make_std_vector_async(d_unique_col_ids, stream);
+  auto column_categories =
+    cudf::detail::make_std_vector_async(d_column_tree.node_categories, stream);
+  auto column_parent_ids =
+    cudf::detail::make_std_vector_async(d_column_tree.parent_node_ids, stream);
+  auto column_range_beg =
+    cudf::detail::make_std_vector_async(d_column_tree.node_range_begin, stream);
+  auto max_row_offsets = cudf::detail::make_std_vector_async(d_max_row_offsets, stream);
+  std::vector<std::string> column_names = copy_strings_to_host(
+    input, d_column_tree.node_range_begin, d_column_tree.node_range_end, stream);
+
+  auto to_json_col_type = [](auto category) {
+    switch (category) {
+      case NC_STRUCT: return json_col_t::StructColumn;
+      case NC_LIST: return json_col_t::ListColumn;
+      case NC_STR:
+      case NC_VAL: return json_col_t::StringColumn;
+      default: return json_col_t::Unknown;
+    }
+  };
+  auto init_to_zero = [stream](auto& v) {
+    thrust::uninitialized_fill(rmm::exec_policy(stream), v.begin(), v.end(), 0);
+  };
+
+  auto initialize_json_columns = [&](auto i, auto& col) {
+    if (column_categories[i] == NC_ERR || column_categories[i] == NC_FN) {
+      return;
+    } else if (column_categories[i] == NC_VAL || column_categories[i] == NC_STR) {
+      col.string_offsets.resize(max_row_offsets[i] + 1, stream);
+      col.string_lengths.resize(max_row_offsets[i] + 1, stream);
+      init_to_zero(col.string_offsets);
+      init_to_zero(col.string_lengths);
+    } else if (column_categories[i] == NC_LIST) {
+      col.child_offsets.resize(max_row_offsets[i] + 2, stream);
+      init_to_zero(col.child_offsets);
```

Comment:
For `d_json_column`, it is not identical. 
child_offsets is identical to cudf list offset. (that's why it is moved instead of copied during cudf column creation).
But string_offset is not similar to cudf string column offsets, it is offset w.r.t _input JSON string_.